truncated_MonteCarlo.py

```diff
@@ -0,0 +1,96 @@
+
+@author: Thomas-Galtier
+"""
+ 
+import numpy as np 
+import scenario
+
+from scipy.stats import norm
+
+
+def truncated_MC(preprocessed_node_list, sv_accuracy=0.001, contrib_accuracy=0.01):
+    """Return the vector of approximated shapeley value corresponding to a list of node and a characteristic function using the truncated monte-carlo method."""
+    
+    characteristic_func=fl_training.compute_test_score
+    n = len(preprocessed_node_list)
+    
+    # we store the value of the characteristic function in order to avoid recomputing it twice
+    char_value_dict={():0} # the dictionary that will countain the values
+    listn=np.arange(n)
+    
+    # return the characteristic function of the nodelist associated to the ensemble  permut, without recomputing it if it was already computed
+    def not_twice_characteristic(permut):
+        #sort permut
+        isin=np.repeat(False,n)
+        isin[permut]=True
+        permut=listn[isin]
+        try: #return the characteristic_func(permut) if it was already computed
+            return char_value_dict[tuple(permut)]
+        except KeyError: #characteristic_func(permut) has not been computedyet, so we compute, store, and return characteristic_func(permut)
+            small_node_list = [preprocessed_node_list[i] for i in permut]
+            char_value_dict[tuple(permut)]= characteristic_func(small_node_list)
+            return char_value_dict[tuple(permut)]
+
+        
+    # truncated_MC initialisation
+    if n==1:
+        return np.array([1])
+    else:
+        characteristic_all_node= not_twice_characteristic(np.arange(n)) #characteristic function on all nodes
+        sv = np.zeros(n) #store current estimation of the shapley values 
+        mean_squares_sv = np.zeros(n) # used for variance computation in the condition of the while loop below 
+        permutation=np.zeros(n) # store the current permutation 
+        firsts_are_equal =True
+        t=0
+        while t<50 or t<=norm.ppf(0.975, loc=0, scale=1)**2* np.max(mean_squares_sv-sv**2) /sv_accuracy**2 or firsts_are_equal: #check if the length of the confidence interval  is below the value of sv_accuracy
+            t+=1
+            print(t<50,t<=norm.ppf(0.975, loc=0, scale=1)**2* np.max(mean_squares_sv-sv**2) /sv_accuracy**2,firsts_are_equal)
+            print(t)
+            print(sv)
+            previous_sv = np.copy(sv)#store previous estimation of the shapley values
+            previous_mean_square_sv= np.copy(mean_squares_sv) #store previous value of   the mean_square_sv
+            previous_permutation=np.copy(permutation)# store the previous permutation
+            permutation = np.random.permutation(n) # store the current permutation
+            if firsts_are_equal and t>1:# makes sure we don't stop the loop because the first permutations are the same (very unlikely but can still happen if n is small)
+                firsts_are_equal=   (permutation==previous_permutation).all()
```

Comment:
The `while` loop has an `or` conditions, so this variable is not going to be used before `t` reaches `50`. I think there is something I don't understand, sorry.

Reply:
Yes but assume we have only the two first conditions in the while. If the 50 first permutations drawn are the same then np.max(mean_squares_sv-sv**2)=0 and the second condition makes the loop stop. firsts_are_equal is here to avoid this early stopping case.

Reply:
Oh, understood. I didn't even consider the possibility of the first 50 being the same :)

Reply:
Oh, understood. I didn't even consider the possibility of the first 50 being the same :)